Implementation of min_makespan 2-apprx algo

CyberSrul opened the pull request:
all done, everything works. passes extensive tests

Israel0545488592 committed:
tests and headers

Israel0545488592 committed:
Revert "tests and headers"

This reverts commit b749dbaf2fc6612cf8220ac7a00ee5738933940e.

requirements: not on the master branch

Israel0545488592 committed:
headers and tests for israels work

CyberSrul committed:
secondary layout

CyberSrul committed:
python psedu code for the entire program

CyberSrul committed:
greedy algo works

more doctest
passes very simple tests
probloms with cvxpy

CyberSrul committed:
LP trouble

CyberSrul committed:
main algo works

CyberSrul committed:
most of the work is done, hopefully

CyberSrul committed:
most of the work is done, hopefully

CyberSrul committed:
fixed bugs more logging

CyberSrul committed:
psedu forest promise holds, better apprx

CyberSrul committed:
full doctest

CyberSrul committed:
implementation + tests complete :)

CyberSrul committed:
binary search procidure fix

CyberSrul committed:
more tests

CyberSrul committed:
simplified interface

CyberSrul committed:
adhearing to libraries rules

CyberSrul committed:
most sparse solution to the LP, done right

CyberSrul committed:
following paper

CyberSrul committed:
Merge branch 'master' of https://github.com/CyberSrul/fairpy

CyberSrul committed:
Merge branch 'master' into israels_work

CyberSrul committed:
merging n' cleaning

CyberSrul committed:
different solver

CyberSrul commented:
min makespan 2 apprx algo

erelsgl commented on fairpy/items/min_makespan.py:
```diff
@@ -0,0 +1,459 @@
+# data structures
```
התיעוד צריך להיות בראש הקובץ, הוא צריך לכלול את שם המאמר ואת שם המתכנת, והפורמט שלו צריך להיות כמו בשאר הקבצים בספריה, למשל כאן:

https://github.com/erelsgl/fairpy/blob/master/fairpy/items/approximation_maximin_share.py

erelsgl replied:
שם הקובץ, ושם הפונקציה הראשית, לא צריכים להיות min_makespan כי הפונקציה לא תמיד מוצאת את המינימום. אפשר לקרוא לקובץ ולפונקציה הראשית על-שם מחברי המאמר:
lenstra_shmoys_tardos_2approximation

erelsgl commented on fairpy/items/min_makespan.py:
```diff
@@ -0,0 +1,459 @@
+# data structures
+import numpy as np
+from fairpy import ValuationMatrix, Allocation, AllocationMatrix
+# psedu random functions
```
לא צריך לשים תיעוד לספריות שמייבאים, בפרט כאשר הן מוכרות וידועות.

erelsgl commented on fairpy/items/min_makespan.py:
```diff
@@ -0,0 +1,459 @@
+'''
+
+
+class scedual:
```
הכתיב הנכון הוא schedule. יש לתקן לאורך כל שני הקבצים.

erelsgl commented on fairpy/items/min_makespan.py:
```diff
@@ -0,0 +1,459 @@
+
+" The Interface Method "
+
+def min_makespan(input: ValuationMatrix) -> Allocation:
```
כיוון שזו הפונקציה הראשית, יש לשים אותה בראש הקובץ. ויש לסמן בבירור לאחר מכן, ששאר הפונקציות הן פונקציות פנימיות.

erelsgl commented on fairpy/items/min_makespan.py:
```diff
@@ -0,0 +1,459 @@
+        Agent #0 gets { 100.0% of 1} with value 2.
+        Agent #1 gets { 100.0% of 2} with value 1.
+        Agent #2 gets { 100.0% of 0} with value 2.
+        <BLANKLINE>
```
יש להוסיף דוגמה של קריאה לפונקציה בעזרת fairpy.divide, כמו שמוסבר ברידמי של הספריה.

erelsgl commented on fairpy/items/min_makespan_demo.py:
```diff
@@ -0,0 +1,169 @@
+from fairpy.items.min_makespan import *
```
הקובץ ששמו מסתיים ב _demo אמור לכלול דוגמאות פשוטות של קריאה לפונקציה הראשית בעזרת fairpy.divide, כמו שמוסבר ברידמי של הספריה. לדוגמה, כמו כאן:
https://github.com/erelsgl/fairpy/blob/master/fairpy/items/propm_allocation_demo.py

הבדיקות המקיפות והחשובות שבקובץ זה אמורות להיות בקובץ ששמו מסתיים ב _test.

erelsgl commented on fairpy/items/min_makespan.py:
```diff
@@ -0,0 +1,459 @@
+    algo(output, **kwargs)
+
+
+def RandomTesting(algo: MinMakespanAlgo, iterations: int, **kwargs):
```
המחלקה הזאת לא צריכה להיות כאן, אלא בקובץ השני (של הבדיקות).

erelsgl commented on fairpy/items/min_makespan.py:
```diff
@@ -0,0 +1,459 @@
+        yield scd.makespan
+
+
+def compare(algo1: Tuple[MinMakespanAlgo, Dict[str, Any]], algo2: Tuple[MinMakespanAlgo, Dict[str, Any]], iterations: int) -> Tuple[int, int]:
```
נראה שהפונקציה הזאת נועדה לבדיקות בלבד. אם כן,יש להעביר אותה לקובץ השני.

erelsgl commented on fairpy/items/min_makespan.py:
```diff
@@ -0,0 +1,459 @@
+import numpy as np
+from fairpy import ValuationMatrix, Allocation, AllocationMatrix
+# psedu random functions
+from numpy.random import randint, uniform
```
נראה שהפונקציות האלה נועדו לבדיקות בלבד. אם כן, יש להעביר אותן לקובץ השני.

erelsgl commented on fairpy/items/min_makespan_demo.py:
```diff
@@ -0,0 +1,169 @@
+
+            MinMakespan(apprx, mat, self.scd)
+
+            self.assertTrue(self.scd.makespan <= 2 * optimum)
```
הבדיקה הזאת נכשלה אצלי כמה פעמים. צריך לוודא שכאשר הבדיקה נכשלת, הקלט האקראי נכתב למסך, כך שיהיה אפשר לברר למה זה נכשל.

Israel0545488592 committed:
PR eequests1

Israel0545488592 committed:
algo limits

on small inputs the approximation algo may fail to deliver 2 approximation
by a small amount thats because it promises an arbitrarly close result
to 2-approximation.
however for ALMOST 100% of caces it does work
and it usually acoures with very similar entries

CyberSrul committed:
more tests requirments and regulations

CyberSrul committed:
resolving conflict

CyberSrul committed:
fixing import line for correct wd

CyberSrul committed:
trying to setesgy on line testing

CyberSrul committed:
Merge branch 'erelsgl:master' into master

erelsgl commented on fairpy/items/min_makespan_2apprximation.py:
```diff
@@ -0,0 +1,423 @@
+
+logger = logging.getLogger(__name__)
+
+logging_format = '%(levelname)s - %(message)s'
```
כל השורות הקשורות לתצורת הלוג הגלובלית לא צריכות להיות כאן, אלא רק בתוכנית הראשית.

erelsgl commented on fairpy/items/min_makespan_2apprximation.py:
```diff
@@ -0,0 +1,423 @@
+Find 2 approximation min makespan on unrelated parallel machines.
+Based on:
+
+Jan Karel Lenstra, David, B. Shmoys and Eva Tardo.
```
בדוק שוב את שמות המחברים

erelsgl commented on fairpy/items/min_makespan_2apprximation.py:
```diff
@@ -0,0 +1,423 @@
+
+    ''' 
+        This method returns a 2-approximation of the optimal minimum makespan.
+        Both as the actuall scedualing and the makespan.
```
יש לתקן שגיאות באנגלית - העזר ב grammarly או כלי אחר.

erelsgl commented on fairpy/items/min_makespan_2apprximation_demo.py:
```diff
@@ -0,0 +1,43 @@
+
+print(divide(min_makespan_2apprximation, input))
+
+print('deals well with dummy caces:\n')
```
יש לתקן שגיאות באנגלית - אפשר להעזר בכלי כגון grammarly.

CyberSrul commented:
למה אתה מתכוון בתוכנית ראשית ? הדמו ? הטסטים ?
בדקתי את שמות המחברים של המאמר, לא הבנתי איפה הטעות אלא אם כן התכוונת לפסיק
ששמתי במקום הלא נכון

On Sun, Jan 29, 2023 at 10:28 PM Erel Segal-Halevi ***@***.***>
wrote:

> ***@***.**** requested changes on this pull request.
> ------------------------------
>
> In fairpy/items/min_makespan_2apprximation.py
> <https://github.com/erelsgl/fairpy/pull/25#discussion_r1090046087>:
>
> > +
>
> +from fairpy import ValuationMatrix, Allocation, AllocationMatrix, divide
>
> +
>
> +import numpy as np
>
> +from numpy.random import randint, uniform
>
> +import cvxpy as cp
>
> +import networkx as nx
>
> +
>
> +from typing import Callable, Iterator, Tuple
>
> +from itertools import product
>
> +from copy import copy
>
> +import logging
>
> +
>
> +logger = logging.getLogger(__name__)
>
> +
>
> +logging_format = '%(levelname)s - %(message)s'
>
>
> כל השורות הקשורות לתצורת הלוג הגלובלית לא צריכות להיות כאן, אלא רק בתוכנית
> הראשית.
> ------------------------------
>
> In fairpy/items/min_makespan_2apprximation.py
> <https://github.com/erelsgl/fairpy/pull/25#discussion_r1090046124>:
>
> > @@ -0,0 +1,423 @@
>
> +'''
>
> +Find 2 approximation min makespan on unrelated parallel machines.
>
> +Based on:
>
> +
>
> +Jan Karel Lenstra, David, B. Shmoys and Eva Tardo.
>
>
> בדוק שוב את שמות המחברים
> ------------------------------
>
> In fairpy/items/min_makespan_2apprximation.py
> <https://github.com/erelsgl/fairpy/pull/25#discussion_r1090046157>:
>
> > +
>
> +stream_handler = logging.StreamHandler()
>
> +stream_handler.setLevel(logging.ERROR)
>
> +formatter = logging.Formatter(logging_format)
>
> +stream_handler.setFormatter(formatter)
>
> +logger.addHandler(stream_handler)
>
> +
>
> +
>
> +
>
> +" The Main Method "
>
> +
>
> +def min_makespan_2apprximation(input: ValuationMatrix) -> Allocation:
>
> +
>
> +    '''
>
> +        This method returns a 2-approximation of the optimal minimum makespan.
>
> +        Both as the actuall scedualing and the makespan.
>
>
> יש לתקן שגיאות באנגלית - העזר ב grammarly או כלי אחר.
> ------------------------------
>
> In fairpy/items/min_makespan_2apprximation_demo.py
> <https://github.com/erelsgl/fairpy/pull/25#discussion_r1090046223>:
>
> > +                        [30, 9, 12]])
>
> +
>
> +print('classic input\n', input, '\nopt: 9\nresult:')
>
> +print(divide(min_makespan_2apprximation, input))
>
> +
>
> +
>
> +input = ValuationMatrix([[6, 2, 4],
>
> +                         [2, 1, 3],
>
> +                         [3, 2, 8]])
>
> +
>
> +print('another input:\n', input, '\nopt: 4\nresult:')
>
> +print(divide(min_makespan_2apprximation, input))
>
> +
>
> +print(divide(min_makespan_2apprximation, input))
>
> +
>
> +print('deals well with dummy caces:\n')
>
>
> יש לתקן שגיאות באנגלית - אפשר להעזר בכלי כגון grammarly.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/erelsgl/fairpy/pull/25#pullrequestreview-1274293138>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/A4RWNJZKNKTHMYS26A5AXQTWU3HFZANCNFSM6AAAAAATWKSMCU>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

erelsgl commented:
"תוכנית ראשית" זה גם הדמו, וגם הקוד שמופיע אחרי 
```
if __name__=="__main__":
```

erelsgl commented:
יש טעות גם בכותרת של המאמר ובקישור למאמר. הקישור הנכון הוא: https://link.springer.com/article/10.1007/BF01585745
ושם המחברת האחרונה הוא Eva **Tardos**.

CyberSrul committed:
polishing

CyberSrul commented:
OK, I commited the last requests
